clang/lib/Sema/SemaConcept.cpp

```diff
@@ -1109,12 +1109,50 @@ bool Sema::EnsureTemplateArgumentListConstraints(
   return false;
 }
 
-bool Sema::CheckInstantiatedFunctionTemplateConstraints(
+static bool CheckFunctionConstraintsWithoutInstantiation(
+    Sema &SemaRef, SourceLocation PointOfInstantiation,
+    FunctionTemplateDecl *Template, ArrayRef<TemplateArgument> TemplateArgs,
+    ConstraintSatisfaction &Satisfaction) {
+  SmallVector<const Expr *, 3> TemplateAC;
+  Template->getAssociatedConstraints(TemplateAC);
+  if (TemplateAC.empty()) {
+    Satisfaction.IsSatisfied = true;
+    return false;
+  }
+
+  LocalInstantiationScope Scope(SemaRef);
+
+  FunctionDecl *FD = Template->getTemplatedDecl();
+  // Collect the list of template arguments relative to the 'primary'
+  // template. We need the entire list, since the constraint is completely
+  // uninstantiated at this point.
+  MultiLevelTemplateArgumentList MLTAL =
+      SemaRef.getTemplateInstantiationArgs(FD, /*DC=*/nullptr,
+                                           /*Final=*/false,
+                                           /*Innermost=*/TemplateArgs,
+                                           /*RelativeToPrimary=*/true,
+                                           /*ForConstraintInstantiation=*/true);
+
+  std::optional<Sema::CXXThisScopeRAII> ThisScope;
+  if (auto *Method = dyn_cast<CXXMethodDecl>(FD))
+    ThisScope.emplace(SemaRef, /*Record=*/Method->getParent(),
+                      /*ThisQuals=*/Method->getMethodQualifiers());
```

Comment:
When is that useful?

Reply:
Honestly, I don't know. :(

I copied that from the previous implementation, and I was expecting it to make a difference in situations where member function calls are involved in a constraint evaluation. But I ran the regression tests with that removed and everything seemed fine.

I will have another try on a build without this patch. It could also be the case we're missing some tests, but I don't know...

Reply:
Huh, I removed these `CXXThisScopeRAII` lines in `CheckInstantiatedFunctionTemplateConstraints()` and all libcxx and clang regressions still passed. I guess we probably could remove these setups there

Reply:
maybe try to play with things like `requires(__is_same_as(s, decltype(*this)));`

Reply:
I tried to contrive a case like

```cpp
template <class T>
struct S {
  void foo() requires(__is_same_as(S&, decltype(*this))) {}
};

void g() {
  S<int>().foo();
}
```

But there would still be no difference if we did the `CXXThisScopeRAII` things away. (It always compiles.)

Maybe I'm running out of ideas right now, so let's just preserve these "strange but plausible" things?